### PR TITLE
add tests around throw cases

### PR DESCRIPTION
**Pull request by mixmix**
This PR:
- adds tests which affirms current throw cases
- adds a couple of cases which should throw (see comments)

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -29,30 +29,55 @@ const NAMED_TYPES = convertTypesToNamedTypes(TYPES)
...
+  MSG.formats['ssb/bendy-butt'].code,
+])
 
 const BLOB = NAMED_TYPES['blob']
 const BLOB_T = Buffer.from([BLOB.code])
-const CLASSIC_BLOB_TF = Buffer.from([BLOB.code, BLOB.formats['ssb/classic'].code])
+const CLASSIC_BLOB_TF = Buffer.from([
+  BLOB.code,
+  BLOB.formats['ssb/classic'].code,
+])
```
Comment (mixmix) [inline index.js:62]: this is all `prettier`, sigh

File: index.js [hunk 2]
```diff
@@ -154,8 +179,7 @@ function encode(input) {
     else if (input.startsWith('%')) return encoder.message(input)
     else if (input.startsWith('&')) return encoder.blob(input)
     else if (input.endsWith('.sig.ed25519')) return encoder.signature(input)
-    else if (input.endsWith('.box2') || input.endsWith('.box'))
-      return encoder.box(input)
+    else if (input.match(/\.box\d*$/)) return encoder.box(input)
```
Comment (mixmix) [inline index.js:182]: Generalised this case to let e.g. `asdsadsa=.box40` through. in `encoder.box` this is then not matched by anything, leading to a throw

File: index.js [hunk 3]
```diff
@@ -250,7 +274,10 @@ function decode(input) {
     else if (t.equals(MSG_T)) return decoder.message(input)
     else if (t.equals(BLOB_T)) return decoder.blob(input)
     else if (t.equals(BOX_T)) return decoder.box(input)
-    else if (tf.equals(SIGNATURE_TF)) return decoder.signature(input)
+    else if (t.equals(SIGNATURE_T)) {
```
Comment (mixmix) [inline index.js:277]: I modified this so that we can catch decoding of buffers which are `type: signature`, but have an unknown `format`, and throw.

Otherwise these were passed through as un-decoded buffers
:fire: (breaking change)

File: test/04-signature.js [hunk 4]
```diff
@@ -10,5 +10,17 @@ tape('04 signature type', function (t) {
   t.deepEqual(encoded[0].slice(0, 2), Buffer.from([4, 0]), 'detected signature')
 
   t.deepEqual(bfe.decode(encoded), values, 'properly decoded')
+
+  t.doesNotThrow(
+    () => bfe.encode('ZG9n.sig.bob'),
+    'unknown signature suffix means encoded as a string'
+  )
```
Comment (mixmix) [inline test/04-signature.js:17]: This is an interested edge case - the lack of sigil means we can't be sure if this is Signature or a String!
I was tempted to write a regex which was something like: `/^<base64pattern>\.sig.\.\w+$/`

But that would be a breaking change and we currently only have one signature and this seems unlikely needed atm

File: test/00-feed.js [hunk 5]
```diff
@@ -16,6 +16,13 @@ tape('00 feed type', function (t) {
 
   t.deepEquals(bfe.decode(encoded), values, 'decode works')
 
+  /* unhappy paths */
+  t.throws(() => bfe.encode('@dogFeed'), 'unknown feedId encode throws')
```
Comment (arj03) [inline test/00-feed.js:20]: Great with these tests, I'm thinking we might need to tighten the check that leads to the feeds branch (similar with the message etc.) because it is rather unfortunate that we can't have strings that starts with @ without being a feed.

Comment (mixmix) [reply test/00-feed.js:20]: We can use is-canonical-base64 style code. Recently did this with box test

Prefix @
Suffix \\.\\w+

(it shows strings which get converted into regexp

Comment (mixmix) [reply test/00-feed.js]: Pushed up a commit with tests and a solution to some of that

Comment (mixmix) [reply test/00-feed.js]: https://github.com/ssb-ngi-pointer/ssb-bfe/pull/18/commits/08c19a531ebf19ab48153d0ee887eda2c5cfc215

Comment (arj03) [reply test/00-feed.js]: Great thanks! The functions in util are similar to the ones in [ssb-ref](https://github.com/ssb-js/ssb-ref/blob/33e800db6025d3c77d27383c16063d3b39f4b06f/index.js#L8) but not quite the same because of classical. We should standardize on those once https://github.com/ssb-ngi-pointer/ssb-meta-feed-spec/issues/19 is settled.

**Comment by arj03** [comment]
This looks great to me, will wait and see if @staltz has any input before merging.

File: .gitignore [hunk 6]
```diff
@@ -1,4 +1,5 @@
 node_modules
 pnpm-lock.yaml
+package-lock.json
 .nyc_output
-coverage
\ No newline at end of file
+coveragep
```
Comment (staltz) [inline .gitignore:5]: What's this?

File: index.js (hunk 2 above)
Comment (staltz) [reply index.js:183]: Wait, do you want this code to throw on `.box` or not? Currently it wouldn't, because the regex is `/\.box\d*$/` not `/\.box\d+$/`.

Note, we have strings suffixed with `.box` in the wild. In fact I think ssb-bfe used to handle `.box1` but this was buggy with test input data, it had to be `.box`.

File: util.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+const IsCanonicalBase64 = require('is-canonical-base64')
```
Comment (staltz) [inline util.js:1]: Would it make sense to use ssb-ref here? This code looks a lot like the implementation of ssb-ref

File: .gitignore (hunk 6 above)
Comment (mixmix) [reply .gitignore:5]: Hmm, typo?
There was some coverage thing in the repo?

File: util.js (hunk 7 above)
Comment (mixmix) [reply util.js:1]: Yeah likely we should migrate there. I would like the general checkers like isBoxLike

**Comment by mixmix** [comment]
Couldn't reply to one of your things for some reason on mobile.
Here is my reply ;

This handles
.box,.box1,.box2 etc etc

They encode.box decides what it actually knows how to encode

Eg
.box30 will currently fail

**Comment by staltz** [comment]
> Couldn't reply to one of your things for some reason on mobile.
> Here is my reply ;
> 
> This handles
> .box,.box1,.box2 etc etc

Ok, good! :)

File: .gitignore (hunk 6 above)
Comment (staltz) [reply .gitignore:5]: Okay, just clarifying that it's not intentional

File: index.js (hunk 2 above)
Comment (mixmix) [reply index.js:183]: I don't want it throwing on .box
I guess .box === .box1,  we should clarify in the encoder

**Comment by mixmix** [comment]
Thanks for the review, merging this now
